File: java/aula05-utils/src/main/java/pt/isel/pc/utils/TimeoutHolder.java
package pt.isel.pc.utils;

public class TimeoutHolder {
    private long toExpire;
    private boolean withTimeout;
    public static final long INFINITE = -1;

    public TimeoutHolder(long millis) {
        if (millis == INFINITE) withTimeout = false;
        else {
            toExpire = System.currentTimeMillis() + millis;
            withTimeout = true;
        }
    }
 
    public long remaining() {
        if (!withTimeout) return Long.MAX_VALUE;
        return Math.max(0, toExpire - System.currentTimeMillis());
    }

    public boolean timeout() {
        return remaining() == 0;
    }
}

File: backup/dotnet/SynchUtils/TimeoutHolder.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SynchUtils {
    public class TimeoutHolder {
        private bool withTimeout;
        private int expired;
        public static readonly long INFINITE = -1;

        public TimeoutHolder(int millis) {
            if (millis == INFINITE) { withTimeout = false; }
            else { expired = Environment.TickCount + millis; withTimeout = true; }
        }

        

        public int Remaining {
            get {
                if (!withTimeout) return int.MaxValue;
                return Math.Max(0, expired - Environment.TickCount);
            }

        }

        public bool Timeout {
            get { return Remaining == 0; }
        }
    }
}
